---
output: 
  html_document:
    code_download: false
    toc: true                  # table of content true
    toc_depth: 3               # upto three depths of headings (specified by #, ## and ###)
    toc_float: true
    number_sections: true      # if you want number sections at each table header
    theme: united              # many options for theme, this one is my favorite.
    highlight: tango           # specifies the syntax highlighting style
    css: 'style.css'
params:
  manhattan:  "covid_1_manhattan.png"
  qqplot:  "covid_1_qqplot_ci.png"
  gwascat: "gwascat_subset.csv"
  trait_type: 'binary'
  saige_results: "saige_results_top_n.csv"
  ldsc_log: "None"
title: "`r paste0('Genomics England GWAS Report' , '') `"
author: ""
date: ""
---


<img src="logo.png" width=100 style="position:absolute;top:4px;bottom:4px;right:4px;" />


```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE)
```

```{r echo = FALSE}
source("DTable.R")
source("sanitise.R")

suppressPackageStartupMessages({

library(magrittr)
library(DT)
library(data.table)
  
library(htmltools)
library(stringr)
library(dplyr)
library(kableExtra)
   })
```

# Visualisations

Plots to explore GWAS results

```{r}
figure_number  <- "Figure 1: "
figure_caption <- "Manhattan plot"
knitr::include_graphics(params$manhattan, )
htmltools::tags$figcaption( style = 'caption-side: bottom; text-align: center; font-size: 85%%; color: #71879d',
                            htmltools::em(figure_number),
                            htmltools::em(figure_caption))
```

```{r}
figure_number  <- "Figure 2: " 
figure_caption <- "QQ plot"
knitr::include_graphics(params$qqplot, )
htmltools::tags$figcaption( style = 'caption-side: bottom; text-align: center; font-size: 85%%; color: #71879d',
                            htmltools::em(figure_number), 
                            htmltools::em(figure_caption))
```

# SAIGE output

```{r}
table   <- as.data.frame(data.table::fread(params$saige_results))
colnames(table) <- snakecase::to_snake_case(colnames(table))
if (params$trait_type =='binary'){
reordered_columns <-
c("chr", "pos", "snpid", "p_value", "imputation_info", "n", "beta", "se", "tstat",
 "p_value_na", "is_spa_converge", "var_t", "var_tstar", "allele_1", "allele_2", "ac_allele_2",
"af_allele_2","af_cases", "af_controls", "n_cases", "n_controls", "hom_n_allele_2_cases",
"het_n_allele_2_cases", "hom_n_allele_2_ctrls", "het_n_allele_2_ctrls")
}
if (params$trait_type == 'quantitative'){
reordered_columns <- colnames(table)
}

table <- table[, reordered_columns]
DTable(table)
```

# GWAS catalogue information

```{r}
table   <- as.data.frame(data.table::fread(params$gwascat))
colnames(table) <- snakecase::to_snake_case(colnames(table))
toDisplayNames <- c( "chr_id", "start", "end", "snps","disease_trait", "mapped_gene","snp_gene_ids", "mapped_trait_uri")
toDisplayIndices <- na.omit(match(toDisplayNames, colnames(table)))
reordered_columns <- c( "chr_id",  "start", "end", "snps", "strand",  "mapped_gene", "reported_gene_s", "mapped_trait", "mapped_trait_uri", "snp_id_current", "risk_allele_frequency",  "pubmedid", "first_author", "date", "journal", "link", "study", "disease_trait", "initial_sample_size", "replication_sample_size", "region",  "upstream_gene_id", "downstream_gene_id", "snp_gene_ids", "upstream_gene_distance", "downstream_gene_distance", "strongest_snp_risk_allele",  "merged", "context", "intergenic", "risk_allele_frequency",  "p_value", "pvalue_mlog", "p_value_text", "or_or_beta", "x_95_ci_text",  "platform_snps_passing_qc", "width","chr_pos", "cnv", "study_accession", "genotyping_technology", "seqnames", "date_added_to_catalog")
table <- table[, reordered_columns]
if (nrow(table) == 0) {
  print("No significant variants found with current analysis have been found in GWAS catalogue.")
} else {
  DTable(table)
}
```
